src/os/bluestore/BlueFS.cc

```diff
@@ -2229,6 +2241,143 @@ int64_t BlueFS::_read_random(
   return ret;
 }
 
+void BlueFS::_wal_update_size(FileRef file) {
+  ceph_assert(file->is_wal_read_loaded);
+  uint64_t current_size = file->fnode.size;
+  uint64_t flush_offset = current_size;
+
+  FileReader *h = new FileReader(file, cct->_conf->bluefs_max_prefetch, false, true);
+
+  while(flush_offset < file->fnode.allocated) {
+    // read first part of wal flush
+    bufferlist bl;
+    _read(h, flush_offset, sizeof(File::WALFlush::WALLength), &bl, nullptr);
+    uint64_t flush_length = 0;
+    ceph_assert(bl.length() >= sizeof(File::WALFlush::WALLength));
+
+    flush_length = *((uint64_t*)bl.c_str());
+    dout(20) << __func__ << " flush_length " << flush_length << dendl;
+
+    if (flush_length == 0) {
+      dout(20) << __func__ << " flush end reached " << dendl;
+      break;
+    }
+    uint64_t increase = flush_length+(sizeof(File::WALFlush::WALLength)+sizeof(File::WALFlush::WALMarker));
+    file->fnode.size += increase;
+    file->wal_flushes.push_back({flush_offset, flush_length});
+    flush_offset += increase;
+  }
+}
+
+int64_t BlueFS::_read_wal(
+  FileReader *h,         ///< [in] read from here
+  uint64_t off,          ///< [in] offset
+  size_t len,            ///< [in] this many bytes
+  bufferlist *outbl,     ///< [out] optional: reference the result here
+  char *out)             ///< [out] optional: or copy it here
+{
```

Comment:
I think we should consider making a simplified logic for _wal_read, since it is only accessed in sequential mode:
`2024-05-10T12:11:30.922+0000 7fe5bea18b40 10 bluefs open_for_read db.wal/000271.log (sequential)`

The simplification would be that regular _read would be used.
We would keep:
- wal offset (as rocksdb sees)
- bluefs offset of wal file
- data read from wal file
- location of next envelope within data already read
And on call to _wal_read() we will prefetch from bluefs file to data buffer if needed,
crop beginning of buffer to output *outbl, and update above pointers.

Reply:
> I think we should consider making a simplified logic for _wal_read, since it is only accessed in sequential mode: `2024-05-10T12:11:30.922+0000 7fe5bea18b40 10 bluefs open_for_read db.wal/000271.log (sequential)`
> 
> The simplification would be that regular _read would be used. We would keep:
> 
> * wal offset (as rocksdb sees)
> * bluefs offset of wal file
> * data read from wal file
> * location of next envelope within data already read
>   And on call to _wal_read() we will prefetch from bluefs file to data buffer if needed,
>   crop beginning of buffer to output *outbl, and update above pointers.

I fail to understand the simplification. 
* wal offset is there, named `wal_data_logical_offset`
* bluefs offset is `flush_offset`. I agree this could be another name
* data read from wal file is kept under FileReader, since it's sequential `_read` already does the heavy lifting of prefetching
* We use bufferlists to gather the output so the crop we do is barely expensive, I know this is some extra work that could be optimized away.